src/main/java/htsjdk/samtools/cram/structure/CRAMRecord.java

```diff
@@ -0,0 +1,927 @@
+        }
+
+        if (isSegmentUnmapped()) {
+            samRecord.setCigarString(SAMRecord.NO_ALIGNMENT_CIGAR);
+        } else {
+            samRecord.setCigar(readFeatures.getCigarForReadFeatures(readLength));
+        }
+
+        if (samRecord.getReadPairedFlag()) {
+            samRecord.setMateReferenceIndex(mateReferenceIndex);
+            samRecord.setMateAlignmentStart(mateAlignmentStart > 0 ?
+                            mateAlignmentStart :
+                            SAMRecord.NO_ALIGNMENT_START);
+            samRecord.setMateNegativeStrandFlag(isMateNegativeStrand());
+            samRecord.setMateUnmappedFlag(isMateUnmapped());
+        } else {
+            samRecord.setMateReferenceIndex(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
+            samRecord.setMateAlignmentStart(SAMRecord.NO_ALIGNMENT_START);
+        }
+
+        samRecord.setInferredInsertSize(templateSize);
+        samRecord.setReadBases(readBases);
+        samRecord.setBaseQualities(qualityScores);
+
+        if (tags != null) {
+            for (final ReadTag tag : tags) {
+                samRecord.setAttribute(tag.getKey(), tag.getValue());
+            }
+        }
+
+        if (readGroupID != NO_READGROUP_ID) {
+            final SAMReadGroupRecord readGroupRecord = samFileHeader.getReadGroups().get(readGroupID);
+            samRecord.setAttribute("RG", readGroupRecord.getId());
+        }
+
+        return samRecord;
+    }
+
+    public void assignReadName() {
+        if (readName == null) {
+            readName = Long.toString(getSequentialIndex());
+            if (nextSegment != null) {
+                nextSegment.readName = readName;
+            }
+            if (previousSegment != null) {
+                previousSegment.readName = readName;
+            }
+        }
+    }
+
+    /**
+     * Set this record's state to normalized. When a CRAM record is read from a CRAM stream, it is "raw" in
+     * that the record's read bases, quality scores, and mate graph are not stored directly as part of the
+     * record. Normalization is the process of resolving these values, and is performed at Slice granularity,
+     * across all records in a Slice.
+     * (see {@link Slice#normalizeCRAMRecords(List, CRAMReferenceRegion, SubstitutionMatrix)}).
+     */
+    public void setIsNormalized() { isNormalized = true; }
+
+    /**
+     * When a CRAM record is read from a CRAM stream, it is "raw" in that the record's read bases, quality
+     * scores, and mate graph are not stored directly as part of the record. These values must be resolved
+     * through the separate process of normalization, which is performed at Slice granularity (all records in a
+     * Slice are normalized at the same time).
+     * (see {@link Slice#normalizeCRAMRecords(List, CRAMReferenceRegion, SubstitutionMatrix)}).
+     * @return true if this record is normalized
+     */
+    public boolean getIsNormalized() { return isNormalized; }
```

Comment:
There should probably be a `normalize()` method on this class that handles all of the alterations itself and sets this flag internally. I also think there should perhaps be more checks in various places ensuring that the class has been normalized before it gets used by other code.

Reply:
Once the mate resolution code is repaired, I can try to move closer to that. But in the end, normalization is a slice-level operation that is done on all records in a slice, rather than something you do to a single record.